008-dplyr-retain-and-exclude.md

```diff
@@ -0,0 +1,909 @@
+existing code out there for this to be possible. That said, all dplyr
+documentation would be updated to use `retain()` or `exclude()`.
+
+## Implementation
+
+Try for yourself at:
+
+``` r
+pak::pak("tidyverse/dplyr@feature/retain-exclude")
+```
+
+Note that this implementation is not finished yet, and may have bugs.
+
+## Motivation
+
+### `filter()` ambiguity
+
+The word “filter” has two meanings depending on the context:
+
+- Filter *for* some values you want to keep
+
+- Filter *out* some values you want to drop
+
+This is unfortunate and can make `filter()` hard to teach. We believe
+that the names `retain()` and `exclude()` are much clearer in their
+intent:
+
+``` r
+data <- tibble(
+  account = c(100, 50, 20, 12, 70)
+)
+
+# Is this keeping rows where `account > 20` or dropping them?
+data |> filter(account > 20)
+```
+
+    ## # A tibble: 3 × 1
+    ##   account
+    ##     <dbl>
+    ## 1     100
+    ## 2      50
+    ## 3      70
+
+``` r
+# These are very clear
+data |> retain(account > 20)
+```
+
+    ## # A tibble: 3 × 1
+    ##   account
+    ##     <dbl>
+    ## 1     100
+    ## 2      50
+    ## 3      70
+
+``` r
+data |> exclude(account > 20)
+```
+
+    ## # A tibble: 2 × 1
+    ##   account
+    ##     <dbl>
+    ## 1      20
+    ## 2      12
+
+### Excluding rows using `this & that`
+
+Due to the lack of an explicit “exclude rows” function in dplyr, many
+people turn to `filter()`. But using `filter()` in this way often
+requires replacing `,` separated conditions with `&` separated
+conditions, wrapping the whole thing in parentheses, and then prefixing
+it all with a `!`.
+
+Take a look at this example. Our goal is:
+
+> *Exclude* rows where the patient is deceased *and* the year of death
+> was before 2012.
+
+``` r
+patients <- tibble(
+  deceased = c(0, 1, 0, 1, 1, 0),
+  date = c(2005, 2010, 2013, 2020, 2010, 2000)
+)
+
+patients |>
+  filter(!(deceased & date < 2012))
+```
+
+    ## # A tibble: 4 × 2
+    ##   deceased  date
+    ##      <dbl> <dbl>
+    ## 1        0  2005
+    ## 2        0  2013
+    ## 3        1  2020
+    ## 4        0  2000
+
+Compare that with this proposed usage of `exclude()`:
+
+``` r
+patients |>
+  exclude(deceased, date < 2012)
+```
+
+    ## # A tibble: 4 × 2
+    ##   deceased  date
+    ##      <dbl> <dbl>
+    ## 1        0  2005
+    ## 2        0  2013
+    ## 3        1  2020
+    ## 4        0  2000
+
+Note how we drop:
+
+- The `!`
+- The `()`
+- The `&` (in favor of sticking with `,`)
+
+This results in an `exclude()` statement that precisely matches the
+intent of the original problem statement. In other words, you can “write
+it like you say it”.
+
+In general, the following form of `filter()` can always be written as a
+much simpler `exclude()`:
+
+``` r
+data |> filter(!(this & that & those))
+data |> exclude(this, that, those)
+```
+
+#### Missing value handling
+
+Let’s introduce some missing values to `patients`.
+
+``` r
+patients <- tibble(
```

Comment:
I wonder if introducing a `names` column might make it easier track and talk about what one expects to happen vs. what does happen?